### PR TITLE
Resolved issues #180 and #183 with a solution of @int3l

### PR DESCRIPTION
**Pull request by ClasherKasten**
I've adapted the solution from @int3l to solve the problem with dot-files he proved in issue #183.

### CODE REVIEW

File: tests/features/save_test.py [hunk 1]
```diff
@@ -276,3 +276,31 @@ def test_vim_force_exit(run, tmpdir):
...
+    with run(str(f)) as h:
+        h.press('hello')
+        h.await_text('hello')
+        h.press_and_enter('^X')
+        h.await_text('file is modified - save [yes, no]?')
+        h.press('y')
+        h.await_text('enter filename: ')
+        h.press('Enter')
+        h.await_exit()
+    assert f.read() == 'hello\n'
```
Comment (asottile) [inline tests/features/save_test.py:306]: this test doesn't actually fail before the change -- the dot was unrelated to the failure

File: tests/features/save_test.py [hunk 2]
```diff
@@ -292,15 +294,19 @@ def test_save_and_exit_to_uncreated_directory(run, tmpdir):
...
-        h.press('hello')
-        h.await_text('hello')
-        h.press_and_enter('^X')
-        h.await_text('file is modified - save [yes, no]?')
-        h.press('y')
-        h.await_text('enter filename: ')
-        h.press('Enter')
-        h.await_exit()
-    assert f.read() == 'hello\n'
+@pytest.mark.parametrize('dot_file', ('.f',))
```
Comment (asottile) [inline tests/features/save_test.py:297]: so again, it doesn't matter that the file starts with a `.`

File: tests/features/save_test.py [hunk 3]
```diff
@@ -292,15 +294,19 @@ def test_save_and_exit_to_uncreated_directory(run, tmpdir):
...
+def test_save_and_exit_to_dot_file(run, dot_file):
+    with open(dot_file, 'w+') as f:
```
Comment (asottile) [inline tests/features/save_test.py:299]: this should not write to the execution working directory -- use a filesystem fixture (`tmpdir`) and `as_cwd()` will probably do what you want

File: tests/features/save_test.py [hunk 4]
```diff
@@ -292,15 +294,19 @@ def test_save_and_exit_to_uncreated_directory(run, tmpdir):
...
+            h.press_and_enter('^X')
+            h.await_text('file is modified - save [yes, no]?')
+            h.press('y')
+            h.await_text('enter filename: ')
+            h.press('Enter')
+            h.await_exit()
+        try:
+            assert f.read() == 'hello\n'
+        finally:
+            os.remove(os.path.abspath(dot_file))
```
Comment (asottile) [inline tests/features/save_test.py:312]: this leaves the file around if an assertion fails above which is very undesirable

Comment (ClasherKasten) [reply tests/features/save_test.py:312]: Nope, the file gets deleted no matter if the assertion fails or not. As you can see in the example where i run `pytest tests/features/save_test.py -k "test_save_and_exit_to_dot_file" ; ll` the file seems to get deleted correctly, but correct me if i'm wrong.
dot-file-test

Comment (asottile) [reply tests/features/save_test.py:312]: key word *above* -- you're not protecting the full critical path where this file could be written

File: tests/features/save_test.py [hunk 5]
```diff
@@ -294,19 +294,20 @@ def test_save_and_exit_to_uncreated_directory(run, tmpdir):
...
-            h.await_text('enter filename: ')
-            h.press('Enter')
-            h.await_exit()
-        try:
-            assert f.read() == 'hello\n'
-        finally:
-            os.remove(os.path.abspath(dot_file))
+def test_save_and_exit_with_relative_file(run, tmpdir):
+    current_cwd = os.getcwd()
+    os.chdir(tmpdir)
```
Comment (asottile) [inline tests/features/save_test.py:299]: tmpdir has `.as_cwd()`

File: tests/features/save_test.py [hunk 6]
```diff
@@ -294,19 +294,20 @@ def test_save_and_exit_to_uncreated_directory(run, tmpdir):
...
+        h.press_and_enter('^X')
+        h.await_text('file is modified - save [yes, no]?')
+        h.press('y')
+        h.await_text('enter filename: ')
+        h.press('Enter')
+        h.await_exit()
+    try:
+        assert f.read() == 'hello\n'
+    finally:
+        os.chdir(current_cwd)
```
Comment (asottile) [inline tests/features/save_test.py:313]: this doesn't protect the critical section any error between line 300 and 309 pollutes global process state

File: tests/features/save_test.py (hunk 5 above)
Comment (ClasherKasten) [reply tests/features/save_test.py]: Realised right know that this comes from pytest, didn't know this. Going to fix that.

File: setup.cfg [hunk 7]
```diff
@@ -1,6 +1,6 @@
 [metadata]
 name = babi
-version = 0.0.25
+version = 0.0.26
```
Comment (asottile) [inline setup.cfg:3]: you should not bump versions in packages you don't own

**Review by asottile — COMMENTED** [review]
typically the way you'd do this from a git perspective is:

(1) `git revert ...`
(2) then commit your fix on top of it

that way it's clear what changed and why

if you could rebase to set that up that would be preferable

File: setup.cfg (hunk 7 above)
Comment (ClasherKasten) [reply setup.cfg:3]: i just pulled the changes after the revert and created a new branch, i never changed the version manually.

Comment (asottile) [reply setup.cfg:3]: well whatever you did there was incorrect -- it shows as you comitting the version change: a75312affc66b44f0

**Comment by ClasherKasten** [comment]
Sorry, don't know if this is right. I didn't have worked much with git in the past. But I try to learn it right now.

**Comment by asottile** [comment]
yeah I think what you'll need to do is:

1. `git checkout upstream/master -b your-branch-name`
2. `git revert c4853f26e433f61493ee4e68de9eb3f10fc8f47e`
3. then make another commit with your fixes and new test

**Pull request closed without merging**